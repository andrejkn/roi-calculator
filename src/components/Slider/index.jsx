import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const SliderContainer = styled.div`
  border-radius: 0px;
  border-color: #888;
  margin: 10px 0px;
`;

const LabelContainer = styled.div`
  text-align: center;
`;

const SliderElement = styled.input`
  width: 100%;
`;

export default function Slider({
  value,
  label,
  minValue,
  maxValue,
  step,
  onChange,
}) {
  return (
    <SliderContainer>
      <LabelContainer>
        {label}
      </LabelContainer>
      <SliderElement
        type="range"
        min={minValue}
        max={maxValue}
        step={step}
        value={value}
        onChange={onChange}
      />
    </SliderContainer>
  );
}

Slider.defaultProps = {
  step: 1,
  label: '',
};

Slider.propTypes = {
  onChange: PropTypes.func.isRequired,
  value: PropTypes.number.isRequired,
  minValue: PropTypes.number.isRequired,
  maxValue: PropTypes.number.isRequired,
  step: PropTypes.number,
  label: PropTypes.string,
};
