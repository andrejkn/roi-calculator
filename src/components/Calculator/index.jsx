import React, { useState } from 'react';
import styled from 'styled-components';

import Slider from '../Slider';

const CalculatorContainer = styled.div`
  margin: 0 30%;
  width: 40%;
  border-radius: 10px;
  border-color: #888;
  border-style: solid;
  border-width: 1px;
  font-family: Segoe UI,Roboto,Helvetica Neue,Arial,Noto Sans,sans-serif
`;

const DetailsContainer = styled.div`
  margin: 10px;
`;

const MonthlyPaymentContainer = styled.div`
  font-weight: 600;
  font-size: 1.25rem;
`;

const SectionLabelContainer = styled.div`
  margin-top: 20px;
  font-weight: 500;
  font-size: 0.85rem;
  color: #AAA;
`;

export default function Calculator() {
  const [state, setState] = useState({
    monthlyPayment: 1500,
    initialInvestment: 2100,
    monthlyInvestment: 0,
    monthsRenting: 60,
  });

  const handleSetInitialInvestment = (evt) => setState({
    ...state,
    initialInvestment: Number(evt.target.value),
  });

  const handleSetMonthlyInvestment = (evt) => setState({
    ...state,
    monthlyInvestment: Number(evt.target.value),
  });

  const handleSetMonthsRenting = (evt) => setState({
    ...state,
    monthsRenting: Number(evt.target.value) * 12,
  });

  const ownershipValue = state.initialInvestment
    + (state.monthlyInvestment * state.monthsRenting);

  return (
    <CalculatorContainer>
      <DetailsContainer>
        MONTHLY PAYMENT
        <MonthlyPaymentContainer>
          {`$${state.monthlyPayment.toLocaleString('en')} rent + $ ${state.monthlyInvestment} investment`}
        </MonthlyPaymentContainer>
        <SectionLabelContainer>
          INITIAL INVESTMENT
        </SectionLabelContainer>
        <Slider
          minValue={2100}
          maxValue={12600}
          step={25}
          value={state.initialInvestment}
          label={`$${state.initialInvestment.toLocaleString('en')}`}
          onChange={handleSetInitialInvestment}
        />
        <SectionLabelContainer>
          MONTHLY INVESTMENT
        </SectionLabelContainer>
        <Slider
          minValue={0}
          maxValue={500}
          step={5}
          value={state.monthlyInvestment}
          label={`$${state.monthlyInvestment.toLocaleString('en')}`}
          onChange={handleSetMonthlyInvestment}
        />
        <SectionLabelContainer>
          YEARS WITH UP&UP
        </SectionLabelContainer>
        <Slider
          minValue={1}
          maxValue={10}
          step={1}
          value={state.monthsRenting / 12}
          label={`${(state.monthsRenting / 12)} years`}
          onChange={handleSetMonthsRenting}
        />
        <SectionLabelContainer>
          OWNERSHIP VALUE
        </SectionLabelContainer>
        <div>
          {`$${ownershipValue.toLocaleString('en')}`}
        </div>
      </DetailsContainer>
    </CalculatorContainer>
  );
}
