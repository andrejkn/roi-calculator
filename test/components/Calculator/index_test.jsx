import React from 'react';
import renderer from 'react-test-renderer';

import Calculator from '../../../src/components/Calculator';
import Slider from '../../../src/components/Slider';

let component = null;

describe('Calculator component test', () => {
  beforeEach(() => {
    component = renderer.create(
      <Calculator />,
    );
  });

  it('should match snapshot', () => {
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should trigger onChange on sliders', () => {
    const sliders = component.root.findAllByType(Slider);
    const event = {
      preventDefault() {},
      target: { value: 1000 }
    };
    renderer.act(() => sliders[0].props.onChange(event));
    expect(sliders[0].props.value).toBe(event.target.value);
    expect(sliders[0].props.label).toBe(`$${event.target.value.toLocaleString()}`);

    renderer.act(() => sliders[1].props.onChange(event));
    expect(sliders[1].props.value).toBe(event.target.value);
    expect(sliders[1].props.label).toBe(`$${event.target.value.toLocaleString()}`);

    renderer.act(() => sliders[2].props.onChange(event));
    expect(sliders[2].props.value).toBe(event.target.value);
    expect(sliders[2].props.label).toBe(`${event.target.value} years`);
  });

  it('should have a correct calculation of ownership value', () => {
    const sliders = component.root.findAllByType(Slider);
    const divs = component.root.findAllByType('div');
  
    const ownershipValue = sliders[0].props.value + (sliders[1].props.value * sliders[2].props.value);
    expect(`$${ownershipValue.toLocaleString()}`)
      .toBe(divs[divs.length - 1].children[0]);
  });
});
