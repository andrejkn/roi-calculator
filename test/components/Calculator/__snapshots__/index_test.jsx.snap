// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator component test should match snapshot 1`] = `
<div
  className="sc-bxivhb qwNia"
>
  <div
    className="sc-ifAKCX dkZAhS"
  >
    MONTHLY PAYMENT
    <div
      className="sc-EHOje gUwQEo"
    >
      $1,500 rent + $ 0 investment
    </div>
    <div
      className="sc-bZQynM kkMeLI"
    >
      INITIAL INVESTMENT
    </div>
    <div
      className="sc-bdVaJa kuQLoz"
    >
      <div
        className="sc-bwzfXH gJIwTT"
      >
        $2,100
      </div>
      <input
        className="sc-htpNat daVAfG"
        max={12600}
        min={2100}
        onChange={[Function]}
        step={25}
        type="range"
        value={2100}
      />
    </div>
    <div
      className="sc-bZQynM kkMeLI"
    >
      MONTHLY INVESTMENT
    </div>
    <div
      className="sc-bdVaJa kuQLoz"
    >
      <div
        className="sc-bwzfXH gJIwTT"
      >
        $0
      </div>
      <input
        className="sc-htpNat daVAfG"
        max={500}
        min={0}
        onChange={[Function]}
        step={5}
        type="range"
        value={0}
      />
    </div>
    <div
      className="sc-bZQynM kkMeLI"
    >
      YEARS WITH UP&UP
    </div>
    <div
      className="sc-bdVaJa kuQLoz"
    >
      <div
        className="sc-bwzfXH gJIwTT"
      >
        5 years
      </div>
      <input
        className="sc-htpNat daVAfG"
        max={10}
        min={1}
        onChange={[Function]}
        step={1}
        type="range"
        value={5}
      />
    </div>
    <div
      className="sc-bZQynM kkMeLI"
    >
      OWNERSHIP VALUE
    </div>
    <div>
      $2,100
    </div>
  </div>
</div>
`;
